nucleus/payara-modules/requesttracing-core/src/main/java/fish/payara/nucleus/requesttracing/store/strategy/ReservoirTraceStorageStrategy.java

```diff
@@ -51,66 +52,42 @@
+     * the same probability of being kept or removed.
      *
-     * @param traces the list of traces to test.
-     * @param maxSize the maximum size of the list.
-     * @return the trace that needs removing, or null if no traces need
-     * removing.
-     */
-    @Override
-    public RequestTrace getTraceForRemoval(Collection<RequestTrace> traces, int maxSize) {
-        if (counter < Long.MAX_VALUE) {
-            counter++;
-        }
-
-        // Probability of keeping the new item
-        double probability = (double) maxSize / counter;
-        boolean keepItem = random.nextDouble() < probability;
-
-        // If the probability fails or the list isn't full, nothing should be removed
-        if (!keepItem || traces.size() <= maxSize) {
-            return null;
-        }
-
-        // Pick a random item in the list
-        int itemToReplace = random.nextInt(traces.size());
-        return traces.toArray(new RequestTrace[]{})[itemToReplace];
-    }
-
-    /**
-     * Gets the trace that needs removing. Removes the provided trace if present, or a random trace ala reservoir 
-     * style trace storage. This method assumes that the reservoir style trace storage removal probability has already
-     * been calculated and come out positive.
-     *
-     * @param traces the list of traces to test.
-     * @param maxSize the maximum size of the list.
+     * @param traces        the list of traces to test.
+     * @param maxSize       the maximum size of the list.
      * @param traceToRemove the trace to remove if present
-     * @return the trace that needs removing, or null if no traces need
-     * removing.
+     * @return the trace that needs removing, or null if no traces need removing.
      */
     @Override
     public RequestTrace getTraceForRemoval(Collection<RequestTrace> traces, int maxSize, RequestTrace traceToRemove) {
         // If the list isn't full nothing should be removed
         if (traces.size() <= maxSize) {
             return null;
         }
-        
-        if (traces.contains(traceToRemove)) {
-            return traceToRemove;
-        } else {
-            // Pick a random item in the list
-            int itemToReplace = random.nextInt(traces.size());
-            return traces.toArray(new RequestTrace[0])[itemToReplace];
+        return traceToRemove != null && traces.contains(traceToRemove) 
+                ? traceToRemove 
+                : findRandomTrace(traces);
+    }
+
+    /**
+     * Since this implementation is only dealing with a special case of the Reservoir sampling scenario where there is a
+     * single element more than the maximum size a selection of equal probability is simply to select any of the
+     * elements at random whereby we are left with the elements to keep.
+     */
+    private RequestTrace findRandomTrace(Collection<RequestTrace> traces) {
+        int itemToReplace = random.nextInt(traces.size());
```

Comment:
The "reservoir sampling" for our case where we always have `n < maxSize` or `n == maxSize+1` indeed is just a random number in range 0  to (n-1) as all we need is to decide which element to throw out to get to `maxSize` again. Since all element should have same probability the random number does just that. Good thing: you do not have to take my word for it, there is a test that proves it.